sdk/src/types/block/semantic.rs

```diff
@@ -209,294 +210,279 @@ impl<'a> ValidationContext<'a> {
-            ),
-            Output::Foundry(output) => (
-                output.amount(),
-                0,
-                Some(output.native_tokens()),
-                Some(output.features()),
-            ),
-            Output::Nft(output) => (
-                output.amount(),
-                output.mana(),
-                Some(output.native_tokens()),
-                Some(output.features()),
-            ),
-            Output::Delegation(output) => (output.amount(), 0, None, None),
-        };
-
-        if let Some(unlock_conditions) = created_output.unlock_conditions() {
-            // Check the possibly restricted address-containing conditions
-            let addresses = unlock_conditions
-                .iter()
-                .filter_map(|uc| match uc {
-                    UnlockCondition::Address(uc) => Some(uc.address()),
-                    UnlockCondition::Expiration(uc) => Some(uc.return_address()),
-                    UnlockCondition::StateControllerAddress(uc) => Some(uc.address()),
-                    UnlockCondition::GovernorAddress(uc) => Some(uc.address()),
-                    _ => None,
-                })
-                .filter_map(Address::as_restricted_opt);
-            for address in addresses {
-                if created_output.native_tokens().map(|t| t.len()).unwrap_or_default() > 0
-                    && !address.has_capability(AddressCapabilityFlag::OutputsWithNativeTokens)
-                {
-                    // TODO: add a variant https://github.com/iotaledger/iota-sdk/issues/1430
-                    return Ok(Some(TransactionFailureReason::SemanticValidationFailed));
+                ),
+                Output::Foundry(output) => (
+                    output.amount(),
+                    0,
+                    Some(output.native_tokens()),
+                    Some(output.features()),
+                ),
+                Output::Nft(output) => (
+                    output.amount(),
+                    output.mana(),
+                    Some(output.native_tokens()),
+                    Some(output.features()),
+                ),
+                Output::Delegation(output) => (output.amount(), 0, None, None),
+            };
+
+            if let Some(unlock_conditions) = created_output.unlock_conditions() {
+                // Check the possibly restricted address-containing conditions
+                let addresses = unlock_conditions
+                    .iter()
+                    .filter_map(|uc| match uc {
+                        UnlockCondition::Address(uc) => Some(uc.address()),
+                        UnlockCondition::Expiration(uc) => Some(uc.return_address()),
+                        UnlockCondition::StateControllerAddress(uc) => Some(uc.address()),
+                        UnlockCondition::GovernorAddress(uc) => Some(uc.address()),
+                        _ => None,
+                    })
+                    .filter_map(Address::as_restricted_opt);
```

Comment:
it should be possible to join those two `filter_map`s?

Reply:
That just moves the line up a bit but it's the same, do you think it's better?

Reply:
Actually I'm going to keep it like that. There is an incoming PR where address capabilities check will be made syntactic instead of semantic. I am planning on moving this code directly to the UCs and provide `addresses` and `restricted_addresses` methods so it needs to be split.

Reply:
okidoki